components/automate-deployment/pkg/converge/converger.go

```diff
@@ -127,63 +137,482 @@ func StartConverger(opts ...ConvergerOpt) Converger {
+			nextState, _, err := converger.currentState.ProcessMessage(
+				converger,
+				&timeout{})
+			if err != nil {
+				logrus.WithError(err).Error("error processing timeout event")
+			}
+			converger.setState(nextState)
+		}
+	}
+}
+
+func (converger *converger) setState(s state) {
+	if s.String() != converger.currentState.String() {
+		logrus.WithFields(logrus.Fields{
+			"old_state": converger.currentState,
+			"new_state": s,
+		}).Debug("converger state transition")
+	}
+	converger.currentState = s
+}
+
+// state represents the possible states of our converge
+// loop. Currently, we use a small state machine to represent the
+// possible "waiting" states that we might end up in as the result of
+// a converge request.
+type state interface {
+	// ProcessMessage takes the converger and the most recently
+	// received message. The returned state is the next state of
+	// the fsm. The returned error may be send to the sender of
+	// the message. The retruned bool indicates whether we should
+	// reset the current state timeout before moving to the next
+	// state.
+	ProcessMessage(*converger, message) (state, bool, error)
+	// TimeoutChan returns a channel that should delivery a
+	// message when this state has timed out. If a message is
+	// received on the timeout channel a timeout message will be
+	// sent to the current state.
+	TimeoutChan() <-chan time.Time
+	// StopTimer should correctly stop any timers or resources
+	// associated with the TimeoutChan.
+	StopTimer()
+	// String is a convenience method used for logging. states should
+	// have uniq string descriptors so we can more easily
+	// conditionally log.
+	String() string
+}
+
+// message represents the messages that can be sent to our
+// converger. Messages move the converger from one state to another,
+// possibly triggering some system change.
+type message interface {
+	// SendResponse delivers err back to the original requestor if
+	// the original requestor expects a response.
+	SendResponse(err error)
+	// String is a convenience method used for logging.
+	String() string
+}
+
+// TODO(ssd) 2019-04-21: Should we just panic here? It is a programming error.
+func UnknownMessageError(m message, s state) error {
+	return errors.Errorf("unknown message %s received while in state %s", m, s)
+}
+
+//
+// Messages
+//
+
+// timeout is a message sent to the current state when the current
+// state's timeout has fired. It is generated by the converger and is
+// not sent by the user code so there is no caller to respond to.
+type timeout struct{}
+
+func (t *timeout) SendResponse(_ error) {}
+func (t *timeout) String() string       { return "timeout" }
+
+// convergeRequest is a user-sendable message that instructs the
+// converger to run the converge loop.
+type convergeRequest struct {
+	eventSink    EventSink
+	desiredState DesiredState
+	doneChan     chan error
+}
+
+func (c *convergeRequest) SendResponse(err error) {
+	c.doneChan <- err
+	close(c.doneChan)
```

Comment:
Interesting.  I'm going to need to look a bit at how this done channel is used.  I feel like for "done" channels, it is a pretty common API pattern in the standard library that the library closes the channel since closing the channel is the only way for the "done channel" to always return.  

This might _feel_ different because the "task" is passed in by the user, but the fact that you even have to pass the task in rather than it being returned by the converge call seems like a quirk of the API. Perhaps if we changed around how the "Task" worked it would be more clear from an API perspective that "you only get to read from this chan"